### PR TITLE
Add DataSource as an optional sourceRef for DataVolumes

### PR DESCRIPTION
**Pull request by arnongilboa**
Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:
A `DataSource` is an abstraction that points to the place we'd like to import/clone from. A `DataVolume` can either have an inlined `dataVolume.Spec.Source` as it exists today, or reference a `DataSource` object via `dataVolume.Spec.SourceRef` to perform the same action. The latter allows us to decouple the inlined source from a DataVolume so we can have a static link that points to a DataSource which is mutable.

For more details, view the [design document](https://github.com/kubevirt/community/blob/master/design-proposals/golden-image-delivery-and-update-pipeline.md).

**Special notes for your reviewer**:
- currently only DataSource sourceRef kind is supported
- currently only PVC DataSource source is supported

**Release note**:
```release-note
Add DataSource as an optional source reference for DataVolumes
```

### CODE REVIEW

**Review by mhenriks — COMMENTED** [review]
I wonder if we can simplify the changes to the DataVolume controller.  Currently, there are calls to `getDataVolumeSourcePVC` sprinkled all over the place.  I wonder if we can simply do the following:

1.  The webhook enforces that all DataVolumes have `spec.source` or `spec.sourceRef` set.  Not both.
2. When DataVolume controller reconciles a DV, it checks if `sourceRef` is set.  If so, it gets the sourceRef and populates `spec.Source`.  All other code stays the same except:
3. When the DataVolume controller updates a DV, have to make sure to nil out `spec.source` when `spec.sourceRef` is set.

I also wonder if we want to support DataSources that do not exist.

**Comment by arnongilboa** [comment]
> I wonder if we can simplify the changes to the DataVolume controller. Currently, there are calls to `getDataVolumeSourcePVC` sprinkled all over the place. I wonder if we can simply do the following:
> 
> 1. The webhook enforces that all DataVolumes have `spec.source` or `spec.sourceRef` set.  Not both.

as currently enforced.

> 2. When DataVolume controller reconciles a DV, it checks if `sourceRef` is set.  If so, it gets the sourceRef and populates `spec.Source` with data from the DataSource.  All other code stays the same except:
> 3. When the DataVolume controller updates a DV, have to make sure to nil out `spec.source` when `spec.sourceRef` is set.
> 
Cool, much smarter and cleaner. although the it makes the solution a bit bit hacky and will require a comment or two explaining the trick.

> I also wonder if we want to support DataSources that do not exist.
We currently don't support sources (e.g PVC) that doesn't exist, so If we do it for DataSources won't it complex a bit your last idea?

File: pkg/apiserver/webhooks/datavolume-mutate.go [hunk 1]
```diff
@@ -71,7 +74,20 @@ func (wh *dataVolumeMutatingWebhook) Admit(ar admissionv1.AdmissionReview) *admi
...
 
-	pvcSource := dataVolume.Spec.Source.PVC
+	if dataVolume.Spec.Source != nil {
+		pvcSource = dataVolume.Spec.Source.PVC
+	} else if dataVolume.Spec.SourceRef != nil && dataVolume.Spec.SourceRef.Kind == cdiv1.DataVolumeDataSource {
+		ns := dataVolume.Namespace
+		if dataVolume.Spec.SourceRef.Namespace != nil && *dataVolume.Spec.SourceRef.Namespace != "" {
+			ns = *dataVolume.Spec.SourceRef.Namespace
+		}
+		dataSource, err := wh.cdiClient.CdiV1beta1().DataSources(ns).Get(context.TODO(), dataVolume.Spec.SourceRef.Name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/apiserver/webhooks/datavolume-mutate.go:84]: I'll sure add it, although not sure it's part of the req/design.

File: pkg/controller/datavolume-controller.go [hunk 2]
```diff
@@ -2061,6 +2072,71 @@ func (r *DatavolumeReconciler) newPersistentVolumeClaim(dataVolume *cdiv1.DataVo
 	return pvc, nil
 }
 
+func (r *DatavolumeReconciler) getDataVolumeSourcePVC(dv *cdiv1.DataVolume) (*cdiv1.DataVolumeSourcePVC, error) {
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/datavolume-controller.go:2075]: Right

**Comment by awels** [comment]
I would like to add, we currently ensure on a clone that the source PVC exists (mainly to check volume modes and size requirements) during validation. Since the dataSourceRef will be mainly cloning, we will probably want to do the same.

File: pkg/controller/datavolume-controller.go [hunk 3]
```diff
@@ -507,46 +508,48 @@ func (r *DatavolumeReconciler) isCSIClonePossible(dataVolume *cdiv1.DataVolume,
...
+	dvSourcePvc, _ := r.getDataVolumeSourcePVC(datavolume)
+	if dvSourcePvc == nil {
+		return NoClone, nil
+	}
+
+	// The source and target PVCs share the same Storage Class
+	// get this the same way as for getSnapshotClassForSmartClone(datavolume, pvcSpec)
+	preferredCloneStrategy, err := r.getCloneStrategy(datavolume)
+	if err != nil {
+		return 0, err
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/datavolume-controller.go:520]: Not my code, just un-tabbed this block... but will sure fix that!

**Comment by arnongilboa** [comment]
> I would like to add, we currently ensure on a clone that the source PVC exists (mainly to check volume modes and size requirements) during validation. Since the dataSourceRef will be mainly cloning, we will probably want to do the same.

@mhenriks - considering that, can we still support dataSources that don't exist?

File: pkg/controller/datavolume-controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (brybacki) [reply pkg/controller/datavolume-controller.go:520]: yes mine ;) And it should be `NoClone`, but it is ignored anyway because of err

**Comment by mhenriks** [comment]
> @mhenriks - considering that, can we still support dataSources that don't exist?

Okay let's make sure DataSources exist for now.

**Comment by mhenriks** [comment]
> Cool, much smarter and cleaner. although the it makes the solution a bit bit hacky and will require a comment or two explaining the trick.

Kind of a hack, but we really should enable the DataVolume status subresource and take away DataVolume controller permission to update DataVolume specs

File: pkg/controller/datavolume-controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/datavolume-controller.go:2075]: Function removed

File: pkg/apiserver/webhooks/datavolume-mutate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/apiserver/webhooks/datavolume-mutate.go:84]: Won't add it, follwing our discussion with @awels .

File: pkg/controller/datavolume-controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/datavolume-controller.go]: fixed

File: tests/api_validation_test.go [hunk 4]
```diff
@@ -113,6 +113,37 @@ var _ = Describe("[rfe_id:1130][crit:medium][posneg:negative][vendor:cnv-qe@redh
...
+					}
+					return false
+				}, timeout, pollingInterval).Should(BeTrue())
+
+			},
+				table.Entry("[test_id:XXXX]fail with empty kind", "", "test", "test-pvc"),
+				table.Entry("[test_id:XXXX]fail with unsupported kind", "no-such-kind", "test", "test-pvc"),
+				table.Entry("[test_id:XXXX]fail with empty sourceRef name", "DataSource", "test", ""),
+				table.Entry("[test_id:XXXX]fail with sourceRef DataSource doesn't exist", "DataSource", "test", "test-pvc"),
+			)
```
Comment (duyanyan) [inline tests/api_validation_test.go:145]: table.Entry("[test_id:6780]fail with empty kind", "", "test", "test-pvc"),
table.Entry("[test_id:6781]fail with unsupported kind", "no-such-kind", "test", "test-pvc"),
table.Entry("[test_id:6782]fail with empty sourceRef name", "DataSource", "test", ""),
table.Entry("[test_id:6783]fail with sourceRef DataSource doesn't exist", "DataSource", "test", "test-pvc"),

File: tests/cloner_test.go [hunk 5]
```diff
@@ -108,6 +108,34 @@ var _ = Describe("all clone tests", func() {
 			completeClone(f, f.Namespace, targetPvc, diskImagePath, sourceMD5, sourcePvcDiskGroup)
 		})
 
+		It("[test_id:XXXX]Should clone imported data from SourceRef PVC DataSource", func() {
```
Comment (duyanyan) [inline tests/cloner_test.go:111]: It("[test_id:6784]Should clone imported data from SourceRef PVC DataSource", func() {

File: tests/api_validation_test.go (hunk 4 above)
Comment (arnongilboa) [reply tests/api_validation_test.go:145]: Thanks Yan, but you missed:
`[test_id:XXXX]fail with both source and sourceRef`

File: tests/api_validation_test.go [hunk 6]
```diff
@@ -284,13 +314,13 @@ var _ = Describe("[rfe_id:1130][crit:medium][posneg:negative][vendor:cnv-qe@redh
...
 			table.Entry("[test_id:1860]fail with missing http url key", "manifests/dvMissingSourceHttp.yaml", true, "Missing Data volume source", "spec.source.http in body must be of type object"),
 			table.Entry("[test_id:1858]fail with missing datavolume spec", "manifests/dvMissingCompleteSpec.yaml", true, "Missing Data volume source", "missing required field \"spec\" in io.kubevirt.cdi.v1beta1.DataVolume", " invalid: spec: Required value"),
 			// k8s < 1.15 return Required value: name or generateName is required, >= 1.15 return error validating data: unknown object type "nil" in DataVolume.metadata
 			table.Entry("[test_id:1857]fail without datavolume name", "manifests/dvNoName.yaml", true, "Required value: name or generateName is required", "error validating data: unknown object type \"nil\" in DataVolume.metadata"),
 			table.Entry("[test_id:1856]fail without meta data", "manifests/dvNoMetaData.yaml", true, "Required value: name or generateName is required"),
+			table.Entry("[test_id:XXXX]fail with both source and sourceRef", "manifests/dvBothSourceAndSourceRef.yaml", true, "Data volume should have either Source or SourceRef"),
```
Comment (duyanyan) [inline tests/api_validation_test.go:323]: table.Entry("[test_id:6786]fail with both source and sourceRef"

File: pkg/apiserver/webhooks/datavolume-validate.go [hunk 7]
```diff
@@ -263,50 +281,90 @@ func (wh *dataVolumeValidatingWebhook) validateDataVolumeSpec(request *admission
...
+	}
+	ns := namespace
+	if spec.SourceRef.Namespace != nil && *spec.SourceRef.Namespace != "" {
+		ns = spec.SourceRef.Namespace
+	}
+	dataSource, err := wh.cdiClient.CdiV1beta1().DataSources(*ns).Get(context.TODO(), spec.SourceRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return &metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueNotFound,
+			Message: fmt.Sprintf("SourceRef %s/%s/%s doesn't exist, %v", spec.SourceRef.Kind, *ns, spec.SourceRef.Name, err),
```
Comment (mhenriks) [inline pkg/apiserver/webhooks/datavolume-validate.go:328]: Error doesn't necessarily mean the resource does not exist.  Could be a problem communicating with apiserver

File: pkg/apiserver/webhooks/datavolume-mutate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (awels) [reply pkg/apiserver/webhooks/datavolume-mutate.go:84]: Normally I would agree with @mhenriks but we are already checingk if source PVCs exists for clone, and since clone sources are going to be the majority of datasource, we should remain consistent.

File: pkg/controller/datavolume-controller.go [hunk 8]
```diff
@@ -347,6 +347,24 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
 		return reconcile.Result{}, nil
 	}
 
+	// If sourceRef is set, populate spec.Source with data from the DataSource
+	if datavolume.Spec.SourceRef != nil {
```
Comment (mhenriks) [inline pkg/controller/datavolume-controller.go:351]: would it be better to extract this to a function?  That reconcile function is already big may be more readable

**Review by mhenriks — COMMENTED** [review]
minor comments.  almost there!

File: pkg/apiserver/webhooks/datavolume-mutate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kwiesmueller) [reply pkg/apiserver/webhooks/datavolume-mutate.go:84]: Is this Get using a watch cache?
Depending on how often we think we'll see these cases it might be better to not cause api requests for this if we can have a watcher instead. Not sure how CDI does that right now.

File: pkg/apiserver/webhooks/datavolume-mutate.go [hunk 9]
```diff
@@ -71,7 +74,20 @@ func (wh *dataVolumeMutatingWebhook) Admit(ar admissionv1.AdmissionReview) *admi
...
+		if err != nil {
+			return toAdmissionResponseError(err)
+		}
+		pvcSource = dataSource.Spec.Source.PVC
```
Comment (kwiesmueller) [inline pkg/apiserver/webhooks/datavolume-mutate.go:88]: Isn't it possible, that a DataSource doesn't point to a PVC yet?
Will this cause any problems here?

Comment (mhenriks) [reply pkg/apiserver/webhooks/datavolume-mutate.go:88]: Yeah should make sure that's not nil

Comment (arnongilboa) [reply pkg/apiserver/webhooks/datavolume-mutate.go:88]: WRT nil PVC, I don't see any difference from the original one a few lines above:
```
if dataVolume.Spec.Source != nil {
    pvcSource = dataVolume.Spec.Source.PVC
} ...
```
Anyway we have a check a few lines below:
```
if pvcSource == nil {
    klog.V(3).Infof("DataVolume %s/%s not cloning", targetNamespace, targetName)
    return allowedAdmissionResponse()
}
```

File: pkg/controller/datavolume-controller.go (hunk 8 above)
Comment (arnongilboa) [reply pkg/controller/datavolume-controller.go:351]: sure

File: tests/manifests/dvMissingSource.yaml [hunk 10]
```diff
@@ -4,6 +4,7 @@ metadata:
   name: test-dv
 spec:
   source:
+  sourceRef:
```
Comment (awels) [inline tests/manifests/dvMissingSource.yaml:7]: why is this needed? it not being there is the same as leaving it blank no?

**Review by awels — COMMENTED** [review]
I am missing something, I see the operator give permissions to the api server to get/list datasources, but I don't see anything giving the operator those permissions, but something must give it those permissions otherwise it can't grant them to the api server.

**Comment by mhenriks** [comment]
@awels the operator's cluster role is constructed as the aggregate of all other roles plus other stuff

File: pkg/apiserver/webhooks/datavolume-mutate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/apiserver/webhooks/datavolume-mutate.go:84]: Not cached but I don't think we expect high volume.  At least initially

File: pkg/apiserver/webhooks/datavolume-validate.go (hunk 7 above)
Comment (mhenriks) [reply pkg/apiserver/webhooks/datavolume-validate.go]: Not a huge deal, just thinking that if returning `CauseTypeFieldValueNotFound` then explicitly check for not found and return more generic error if something else fails

**Comment by awels** [comment]
Gotcha, okay, then that was the part I was missing. Ignore that comment then.

File: tests/manifests/dvMissingSource.yaml (hunk 10 above)
Comment (arnongilboa) [reply tests/manifests/dvMissingSource.yaml:7]: No real reason. Just because it was there already for `source`.
So may we remove this yaml together with `test_id:1762` and keep only `test_id:1861`. They are simillar except the blank pvc.
```
$ diff tests/manifests/dvMissingSource.yaml tests/manifests/dvMissingSource2.yaml
6,7d5
<   source:
<   sourceRef:
```

File: pkg/apiserver/webhooks/datavolume-mutate_test.go [hunk 11]
```diff
@@ -180,6 +181,7 @@ func mutateDVs(key *rsa.PrivateKey, ar *admissionv1.AdmissionReview, isAuthorize
 		}
 		return true, sar, nil
 	})
-	wh := NewDataVolumeMutatingWebhook(client, key)
+	cdiClient := cdiclientfake.NewSimpleClientset()
+	wh := NewDataVolumeMutatingWebhook(client, cdiClient, key)
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/apiserver/webhooks/datavolume-mutate_test.go:254]: added
